content/blog/fast-data-processing-pipeline-for-predicting-flight-delays-using-apache-apis-kafka-spark-streaming-and-machine-learning-part-3.md

```diff
@@ -0,0 +1,295 @@
+
+![JSON Schema Flexibility](/img/json-schema-flexibility.png)
+
+JSON facilitates the natural evolution of your data schema during the life of your application. For example, suppose at first we have the following schema, where each JSON message has the predicted flight delay using a decision tree:
+
+```json
+{
+    "_id": "UA_2017-03-28_DEN_SFO_721",
+    "dofW": 2,
+    "carrier": "UA",
+    "origin": "DEN",
+    "dest": "SFO",
+    "crsdephour": 11,
+    "crsdeptime": 1120.0,
+    "crsarrtime": 1308.0,
+    "crselapsedtime": 168.0,
+    "pred_dtree": 1.0
+}
+```
+
+Later, you can easily capture more prediction data values quickly without changing the architecture of your application and without updating a database schema, by adding attributes. In the example below, we have added predictions for other machine learning models. These can be added dynamically to the same document instance in MapR Database without any database schema changes:
+
+```json
+{
+    "_id": "UA_2017-03-28_DEN_SFO_721",
+    "dofW": 2,
+    "carrier": "UA",
+    "origin": "DEN",
+    "dest": "SFO",
+    "crsdephour": 11,
+    "crsdeptime": 1120.0,
+    "crsarrtime": 1308.0,
+    "crselapsedtime": 168.0,
+    "pred_dtree": 1.0,
+    "pred_randforest": 1.0,
+    "pred_svm": 1.0,
+    "actual_delay": 1.0
+
+
+}
+```
+
+MapR Event Store allows processing of the same messages by different consumers. This makes it easy to add different consumers for the same message. With this type of architecture and flexible schema, you can easily add and deploy new microservices with new machine learning models.
+
+**Spark Streaming writing to MapR Database**
+
+The MapR Database OJAI Connector for Apache Spark enables you to use [MapR Database as a sink for Apache Spark Data Streams](https://docs.datafabric.hpe.com/62/Spark/SavingDStreamMapRDB.html).
+
+![Spark Streaming writing to MapR Database](/img/spark-streaming-writing-to-mapr-db.png)
+
+You can read about the MapR Event Streams Spark Streaming code in [part 2 of this series](https://developer.hpe.com/blog/fast-data-processing-pipeline-for-predicting-flight-delays-using-apache-apis-kafka-spark-streaming-and-machine-learning-part-2/); here, we will focus on Spark streaming writing to MapR Database. The messages from the MapR Database topic are in JSON format and contain the following for each flight: the flight id, day of the week, carrier, origin, destination, scheduled departure hour, scheduled departure time, scheduled arrival time, scheduled travel time, delay prediction, and actual delay label (Note in the real world architecture the actual delay label would come later in a different topic, but to keep the tutorial code simple it is combined here). An example is shown below:
```

Comment:
Add hyphen between “real” and “world”
Add commas after “architecture” and “simple”

![image](https://user-images.githubusercontent.com/77496261/137183305-fc145a6f-5a03-48e9-a02c-1f76ec78e007.png)

Reply:
done